framework/include/utils/CellCenteredMapFunctor.h

```diff
@@ -0,0 +1,151 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include "MooseFunctor.h"
+#include "GreenGaussGradient.h"
+#include "MathFVUtils.h"
+#include "libmesh/utility.h"
+#include "libmesh/type_tensor.h"
+#include "libmesh/compare_types.h"
+#include "libmesh/threads.h"
+
+template <typename T, typename T2, typename std::enable_if<ScalarTraits<T>::value, int>::type = 0>
+inline TypeVector<typename CompareTypes<T, T2>::supertype>
+outer_product(const T & a, const TypeVector<T2> & b)
+{
+  TypeVector<typename CompareTypes<T, T2>::supertype> ret;
+  for (unsigned int i = 0; i < LIBMESH_DIM; i++)
+    ret(i) = a * b(i);
+
+  return ret;
+}
+
+/**
+ * A functor whose evaluation relies on querying a map where the keys are element ids and the values
+ * correspond to the element/cell values. This is a very useful data type for storing the result of
+ * (possibly repeated) interpolation and reconstruction
+ */
+template <typename T, typename Map>
+class CellCenteredMapFunctor : public Moose::FunctorBase<T>, public Map
+{
+public:
+  using typename Moose::FunctorBase<T>::ValueType;
+  using typename Moose::FunctorBase<T>::GradientType;
+  using typename Moose::FunctorBase<T>::DotType;
+  using ElemArg = Moose::ElemArg;
+  using ElemFromFaceArg = Moose::ElemFromFaceArg;
+  using FaceArg = Moose::FaceArg;
+  using SingleSidedFaceArg = Moose::SingleSidedFaceArg;
+  using ElemQpArg = Moose::ElemQpArg;
+  using ElemSideQpArg = Moose::ElemSideQpArg;
+
+  CellCenteredMapFunctor(const MooseMesh & mesh) : _mesh(mesh) {}
+
+  bool isExtrapolatedBoundaryFace(const FaceInfo & fi) const override { return !fi.neighborPtr(); }
+
+private:
+  const MooseMesh & _mesh;
+
+  ValueType evaluate(const ElemArg & elem_arg, unsigned int) const override final
+  {
+    const Elem * const elem = elem_arg.elem;
+    return libmesh_map_find(*this, elem->id());
```

Comment:
what could cause the search to fail? I m running into this for the msfr input, which later takes out the solver. 
      Computing Residualmap_find() error: key "2649" not found in file /raid/moose/framework/build/header_symlinks/CellCenteredMapFunctor.h on line 60

Reply:
These are at the top of the backtrace on those:
```
INSFVRhieChowInterpolator::computeFirstAndSecondOverBars()
23: INSFVRhieChowInterpolator::finalizeBData()
24: INSFVRhieChowInterpolator::finalize()
```

Reply:
if I delete inactive blocks (no fluid flow in reflector and shield), error goes away. 
The RC UO has been restricted to the fluid flow domain already

Reply:
This would fail if something is trying to read at an element id that we haven't written to. Like if you tried to evaluate the porosity at a domain that the interpolator object doesn't extend to

Reply:
this is the msfr so it's strictly the INSFVRhieChowInterpolator and there's no porosity. 
I think this is the body forces

Reply:
Ok interesting. Can you tell me what input so I can try it out?

Reply:
https://github.com/idaholab/virtual_test_bed/blob/PR_rcuo/msr/msfr/steady/run_ns_initial.i